docs/tutorials/build-files-archive-with-dal.md

```diff
@@ -0,0 +1,689 @@
+
+    debug_msg!(host, "Saving slot under `{}'\n", path);
+
+    let () = host.store_write_all(&path, &buffer)?;
+
+    Ok(())
+}
+```
+
+There is a bit to unpack here.
+
+1. First, we allocate a buffer of the size of a slot, not a size of the page.
+2. Second, we try to fetch the contents of the first page. If 0 bytes are
+   written by `reveal_dal_page`, then we know the targeted slot has not been
+   attested for this block, and we have nothing left to do.
+3. Otherwise, we read as many page as necessary to get the full slot.
+4. Once it’s done, we store said slot in the durable storage, using the Blake2B
+   hash (encoded in hexadecimal) as its key.
+
+We use two additional dependencies for this: `tezos_crypto_rs` for hashing, and
+`hex` for encoding.
+
+```diff
++use tezos_crypto_rs::blake2b;
++use tezos_smart_rollup::storage::path::OwnedPath;
+ use tezos_smart_rollup::{host::RuntimeError, kernel_entry, prelude::*};
+ use tezos_smart_rollup_host::dal_parameters::RollupDalParameters;
+```
+
+```diff
+ [dependencies]
+ tezos-smart-rollup = { version = "0.2.2", features = [ "proto-alpha" ] }
+ tezos-smart-rollup-host = { version = "0.2.2", features = [ "proto-alpha" ] }
++tezos_crypto_rs = { version = "0.5.2", default-features = false }
++hex = "0.4.3"
+```
+
+:::note
+Adding `default-features = false` for `tezos_crypto_rs` is necessary for the
+crate to be compatible with Smart Rollups.
+:::
+
+Deploy your Smart Rollup again, publish a file, wait for enough levels, and you
+should see a line of log giving you the hash used to register the slot.
+
+Assuming you have set the `hash` variable accordingly, then you can check the
+file indeed exists (and later use `hexdump -C` to inspect its contents.
+
+```bash
+curl "http://localhost:8932/global/block/head/durable/wasm_2_0_0/value?key=/${hash}" \
+    -H 'Content-Type: application/octet-stream'
+    -o slot.bin
+```
+
+:::note Why `diff` won’t work
+You cannot use `diff` to ensure the file you originally published and the one
+you downloaded from the rollup node are equal. Indeed, they are not: since
+the size of a slot is fixed, the DAL node pads the value it receives from
+`POST /slot`. in order to ensure it has the correct size.
+:::
+
+## Conclusion
+
+Our files archive is now completed. At least, the features we wanted to
+implement are there: a Smart Rollup storing in its durable storage the slots
+attested at a given index.
+
+From there, the sky’s the limit. More features could be implemented, like
+having the files publishers pay for the storage they are using in Layer 2 (by
+allowing them to deposit XTZ to the Smart Rollup, and sign the files they
+publish). We could also build a frontend to visualize the files published in
+our archive. Or, we could deal with the fact that for now, it is not possible
+for a consumer of the file to know it’s original size (we could fix that by
+modifying the script we use to publish a file to prefix it with its size).
+
```

Comment:
Maybe also mention some of the limitations of the archive: it handles at most one file per level and the file size cannot exceed the DAL slot size.

Reply:
https://github.com/trilitech/tezos-developer-docs/pull/264/commits/c25183f5ccac64203cee88a0480898b5738eae4a